src/libpgmoneta/walfile/wal_reader.c

```diff
@@ -817,3 +814,270 @@ pgmoneta_wal_get_record_block_tag_extended(struct decoded_xlog_record* pRecord,
    }
    return true;
 }
+
+void
+pgmoneta_wal_record_display(struct decoded_xlog_record* record, uint8_t magic_value, enum value_type type)
+{
+   if (type == ValueJSON)
+   {
+      struct value* record_serialized = NULL;
+      char* value_str = NULL;
+      pgmoneta_wal_record_serialize(record, magic_value, ValueJSON, &record_serialized);
+      value_str = pgmoneta_value_to_string(record_serialized, FORMAT_JSON_COMPACT, NULL, 0);
+      printf("%s\n", value_str);
+      pgmoneta_value_destroy(record_serialized);
+      free(value_str);
+   }
+   else if (type == ValueString)
+   {
+      char* header_str = NULL;
+      char* rm_desc = NULL;
+      char* backup_str = NULL;
+      char* prev_lsn_string = NULL;
+      uint32_t rec_len;
+      uint32_t fpi_len;
+
+      pgmoneta_wal_get_record_length(record, &rec_len, &fpi_len);
+      prev_lsn_string = pgmoneta_lsn_to_string(record->header.xl_prev);
+
+      header_str = pgmoneta_format_and_append(header_str, "%s | %d | %d | %d | %s",
+                                              RmgrTable[record->header.xl_rmid].name, rec_len, record->header.xl_tot_len, record->header.xl_xid, prev_lsn_string);
+
+      rm_desc = RmgrTable[record->header.xl_rmid].rm_desc(rm_desc, record);
+
+      backup_str = pgmoneta_wal_get_record_block_ref_info(backup_str, record, false, true, &fpi_len, magic_value);
+
+      printf("%s%s%s | %s%s%s %s%s%s\n",
+             COLOR_RED, header_str, COLOR_RESET,
+             COLOR_GREEN, rm_desc, COLOR_RESET,
+             COLOR_BLUE, backup_str, COLOR_RESET);
+      free(header_str);
+      free(rm_desc);
+      free(backup_str);
+      free(prev_lsn_string);
+   }
```

Comment:
It is a lot easier to read if you do the declarations first, and always end with `free(...); return;`

We can discuss the `ValueString` formatting later... f.ex. it is important that `RmgrTable[record->header.xl_rmid].name` has its own color

Reply:
I've put the definitions at the beginning. 
since there are two different paths and to avoid having a lot of
`if (x == NULL) free(x);` at the end, I've put the freeing logic at the end of each branch